src/System.Management.Automation/engine/remoting/fanin/OutOfProcTransportManager.cs

```diff
@@ -663,29 +682,82 @@ private void OnCloseSessionCompleted()
         {
             // stop timer
             _closeTimeOutTimer.Change(Timeout.Infinite, Timeout.Infinite);
+
+            // Stop protocol message processing threads.
+            _sessionMessageQueue.CompleteAdding();
+            _commandMessageQueue.CompleteAdding();
+
             RaiseCloseCompleted();
             CleanupConnection();
         }
 
         protected abstract void CleanupConnection();
 
+        private void ProcessMessageProc(object state)
+        {
+            var messageQueue = state as BlockingCollection<string>;
+
+            try
+            {
+                while (true)
+                {
+                    var data = messageQueue.Take();
+                    try
+                    {
+                        OutOfProcessUtils.ProcessData(data, _dataProcessingCallbacks);
+                    }
+                    catch (Exception exception)
+                    {
+                        PSRemotingTransportException psrte =
+                            new PSRemotingTransportException(PSRemotingErrorId.IPCErrorProcessingServerData,
+                                RemotingErrorIdStrings.IPCErrorProcessingServerData,
+                                exception.Message);
+                        RaiseErrorHandler(new TransportErrorOccuredEventArgs(psrte, TransportMethodEnum.ReceiveShellOutputEx));
+                    }
+                }
+            }
+            catch (InvalidOperationException)
+            {
+                // Normal session message processing thread end.
+            }
+        }
+
+        private const string GUIDTAG = "PSGuid='";
+        private const int GUID_STR_LEN = 36;        // GUID string: 32 digits plus 4 dashes
+        private Guid GetMessageGuid(string data)
+        {
+            // Scan data packet for a GUID.
+            var iTag = data.IndexOf(GUIDTAG, StringComparison.OrdinalIgnoreCase);
+            if (iTag > -1)
+            {
+                try
+                {
+                    var psGuidString = data.Substring(iTag + GUIDTAG.Length, GUID_STR_LEN);
+                    return new Guid(psGuidString);
+                }
+                catch { }
```

Comment:
This is just handling the case where `Substring()` fails because the `data` is too short?   Perhaps just catch MethodInvocationException to make this explicit and add a comment.

Reply:
This is to protect against malformed packets.  I don't want to do a full packet decode here because it will be done later during processing, where any malformed issues are reported.  I just want to do a quick scan here so that I can route the message appropriately.

I'll add a comment.